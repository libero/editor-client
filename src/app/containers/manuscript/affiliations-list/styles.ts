import { makeStyles } from '@material-ui/core/styles';

export const useAffiliationStyles = makeStyles((theme) => ({
  addButton: {
    marginLeft: theme.spacing(1)
  },
  listItem: {
    display: 'flex',
    alignItems: 'flex-start',
    lineHeight: '24px',
    marginTop: theme.spacing(1),
    '&:first-child': {
      marginTop: 0
    }
  },
  affiliationInfo: {
    flex: 1
  },
  orderLabel: {
    fontSize: 10,
    lineHeight: '16px',
    paddingRight: 2
  },
  editButton: {
    flexGrow: 0,
    padding: 6,
    marginLeft: theme.spacing(2)
  },
  orcidIcon: {
    width: 16,
    height: 16,
    marginTop: 4,
    display: 'inline',
    verticalAlign: 'top'
  }
}));
