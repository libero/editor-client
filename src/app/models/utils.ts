import { pick, get } from 'lodash';
import { Schema, SchemaSpec } from 'prosemirror-model';

import { nodes } from 'app/models/config/nodes';
import { marks } from 'app/models/config/marks';

export function makeSchemaFromConfig(topNode: string, nodeNames: string[], markNames: string[]): Schema {
  const filteredNodes = pick(nodes, nodeNames);
  const filteredMarks = pick(marks, markNames);

  return new Schema({
    nodes: filteredNodes,
    marks: filteredMarks,
    topNode
  } as SchemaSpec<keyof typeof filteredNodes, keyof typeof filteredMarks>);
}

export function getTextContentFromPath(el: ParentNode, path): string {
  return get(el.querySelector(path), 'textContent');
}
