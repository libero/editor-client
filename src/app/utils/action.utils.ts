import { createAction, SimpleActionCreator } from 'redux-act';

export interface Action<T> {
  type: string;
  payload: T;
}

export interface AsyncAction<R, T> {
  request: SimpleActionCreator<R>;
  success: SimpleActionCreator<T>;
  error: SimpleActionCreator<Error>;
}

export function createAsyncAction<R, T>(type: string): AsyncAction<R, T> {
  return {
    request: createAction<R>(type + '_REQUEST'),
    success: createAction<T>(type + '_SUCCESS'),
    error: createAction<Error>(type + '_ERROR')
  } as AsyncAction<R, T>;
}
