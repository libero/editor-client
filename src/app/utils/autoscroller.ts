const WINDOW_SCROLL_ZONE = 300;

interface Coords {
  x: number;
  y: number;
}

interface ScrollUpdateInterface {
  translate: Coords;
  maxTranslate: Coords;
  minTranslate: Coords;
  minScroll: Coords;
  width: number;
  height: number;
}

export class AutoScroller {
  private static instance: AutoScroller;

  public static getInstance(): AutoScroller {
    if (!AutoScroller.instance) {
      AutoScroller.instance = new AutoScroller();
    }
    return AutoScroller.instance;
  }

  private interval: number | null;
  private container: Element;
  private isAutoScrolling: boolean;

  constructor() {
    this.container = window.document.scrollingElement || window.document.documentElement;
  }

  public clear(): void {
    if (this.interval == null) {
      return;
    }

    clearInterval(this.interval);
    this.interval = null;
  }

  public updateScroll(event: DragEvent, dropContainer: Element): void {
    const containerRect = this.container.getBoundingClientRect();
    const dropRect = dropContainer.getBoundingClientRect();

    const minTranslate: Coords = { x: 0, y: 0 };
    const maxTranslate: Coords = {
      x: containerRect.width,
      y: containerRect.height
    };

    const minScroll = {
      x: dropRect.left + this.container.scrollLeft,
      y: dropRect.top + this.container.scrollTop
    };

    this.doAutoscroll({
      translate: { x: event.clientX, y: event.clientY },
      minTranslate,
      maxTranslate,
      minScroll,
      width: WINDOW_SCROLL_ZONE * 2,
      height: WINDOW_SCROLL_ZONE * 2
    });
  }

  private doAutoscroll({ translate, minTranslate, maxTranslate, width, height }: ScrollUpdateInterface): void {
    const direction = {
      x: 0,
      y: 0
    };

    const speed = {
      x: 1,
      y: 1
    };

    const acceleration = {
      x: 10,
      y: 10
    };

    const { scrollTop, scrollLeft, scrollHeight, scrollWidth, clientHeight, clientWidth } = this.container;

    const isTop = scrollTop === 0;
    const isBottom = scrollHeight - scrollTop - clientHeight === 0;
    const isLeft = scrollLeft === 0;
    const isRight = scrollWidth - scrollLeft - clientWidth === 0;

    if (translate.y >= maxTranslate.y - height / 2 && !isBottom) {
      // Scroll Down
      direction.y = 1;
      speed.y = acceleration.y * Math.abs((maxTranslate.y - height / 2 - translate.y) / height);
    } else if (translate.x >= maxTranslate.x - width / 2 && !isRight) {
      // Scroll Right
      direction.x = 1;
      speed.x = acceleration.x * Math.abs((maxTranslate.x - width / 2 - translate.x) / width);
    } else if (translate.y <= minTranslate.y + height / 2 && !isTop) {
      // Scroll Up
      direction.y = -1;
      speed.y = acceleration.y * Math.abs((translate.y - height / 2 - minTranslate.y) / height);
    } else if (translate.x <= minTranslate.x + width / 2 && !isLeft) {
      // Scroll Left
      direction.x = -1;
      speed.x = acceleration.x * Math.abs((translate.x - width / 2 - minTranslate.x) / width);
    }

    if (this.interval) {
      this.clear();
      this.isAutoScrolling = false;
    }

    if (direction.x !== 0 || direction.y !== 0) {
      this.interval = window.setInterval(() => {
        this.isAutoScrolling = true;
        const offset = {
          left: speed.x * direction.x,
          top: speed.y * direction.y
        };
        this.container.scrollTop += offset.top;
        this.container.scrollLeft += offset.left;
      }, 5);
    }
  }
}
