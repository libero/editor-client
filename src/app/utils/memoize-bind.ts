import { memoize } from 'lodash';
import objectHash from 'object-hash';

export const memoizeBind = memoize(
  (fn: Function, ...args: any[]) => {
    return fn.bind(null, ...args);
  },
  (fn: Function, firstArg: any) => {
    return objectHash({
      fn,
      firstArg
    });
  }
);
