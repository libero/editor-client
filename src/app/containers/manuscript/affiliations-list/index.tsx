import React, { useCallback } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { IconButton } from '@material-ui/core';
import EditIcon from '@material-ui/icons/Edit';

import { SectionContainer } from 'app/components/section-container';
import { ActionButton } from 'app/components/action-button';
import { getAffiliations } from 'app/selectors/manuscript.selectors';
import { Affiliation } from 'app/models/affiliation';
import { useAffiliationStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import { ConnectedAffiliationFormDialog } from 'app/containers/affiliation-form-dialog';
import { ComponentWithId } from 'app/types/utility.types';

export const AffiliationsList: React.FC<ComponentWithId> = ({ id }) => {
  const classes = useAffiliationStyles();
  const affiliations = useSelector(getAffiliations);
  const dispatch = useDispatch();

  const editAffiliation = useCallback(
    (aff: Affiliation) => {
      dispatch(
        manuscriptEditorActions.showModalDialog({
          component: ConnectedAffiliationFormDialog,
          props: { affiliation: aff },
          title: 'Edit Affiliation'
        })
      );
    },
    [dispatch]
  );

  const addAffiliation = useCallback(() => {
    dispatch(
      manuscriptEditorActions.showModalDialog({
        component: ConnectedAffiliationFormDialog,
        props: {},
        title: 'Add Affiliation'
      })
    );
  }, [dispatch]);

  const renderAffiliation = (aff: Affiliation): React.ReactNode => (
    <li key={aff.id} className={classes.listItem}>
      <div className={classes.orderLabel}>{affiliations.length > 1 ? `(${aff.label})` : undefined}</div>
      <div className={classes.affiliationInfo}>{aff.getDisplayName()}</div>
      <IconButton classes={{ root: classes.editButton }} onClick={editAffiliation.bind(null, aff)}>
        <EditIcon fontSize="small" />
      </IconButton>
    </li>
  );

  return (
    <section>
      <SectionContainer label="Affiliations" id={id}>
        <ul className={classes.root}>{affiliations.map(renderAffiliation)}</ul>
      </SectionContainer>
      <ActionButton
        title="Add Affiliation"
        variant="addEntity"
        onClick={addAffiliation}
        className={classes.addButton}
      />
    </section>
  );
};
