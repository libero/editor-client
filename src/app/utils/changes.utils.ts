import { cloneDeepWith } from 'lodash';
import { EditorState } from 'prosemirror-state';

import { JSONObject } from 'app/types/utility.types';
import { Manuscript } from 'app/types/manuscript';
import { Change } from 'app/utils/history/change';
import { BatchChange } from 'app/utils/history/batch-change';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';
import { RearrangingChange } from 'app/utils/history/rearranging-change';
import { UpdateObjectChange } from 'app/utils/history/update-object-change';
import { DeleteObjectChange } from 'app/utils/history/delete-object-change';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { Affiliation } from 'app/models/affiliation';
import { Person } from 'app/models/person';
import { Reference } from 'app/models/reference';
import { RelatedArticle } from 'app/models/related-article';
import { Keyword } from 'app/models/keyword';
import { AddObjectChange } from 'app/utils/history/add-object-change';

export function manuscriptEntityToJson<T>(object: T): JSONObject {
  return cloneDeepWith(object, (value) => {
    if (value instanceof EditorState) {
      return value.toJSON();
    }
  });
}

export function deserializeChanges(changesJson: JSONObject[]): Change[] {
  return changesJson.map((changeData) => {
    switch (changeData.type) {
      case 'prosemirror':
        return ProsemirrorChange.fromJSON(changeData);

      case 'rearranging':
        return RearrangingChange.fromJSON(changeData);

      case 'batch':
        return BatchChange.fromJSON(changeData);

      case 'update-object':
        return UpdateObjectChange.fromJSON(changeData);

      case 'delete-object':
        return DeleteObjectChange.fromJSON(changeData);

      case 'add-object':
        return AddObjectChange.fromJSON(changeData);

      default:
        throw new TypeError(
          `Cannot deserialize ${changeData.type} change. Change type invalid, JSON object malformed or handling not supported.`
        );
    }
  });
}

export function deserializeBackmatter(path: string, json: JSONObject): BackmatterEntity {
  if (path.indexOf('affiliations') >= 0) {
    return new Affiliation(json);
  }

  if (path.indexOf('authors') >= 0) {
    return new Person(json);
  }

  if (path.indexOf('references') >= 0) {
    return new Reference(json);
  }

  if (path.indexOf('relatedArticles') >= 0) {
    return new RelatedArticle(json);
  }

  if (path.match(/keywordGroups\.[^.]+\.keywords/)) {
    return new Keyword(json);
  }

  throw new Error(`deserialization of backmatter entity for  ${path} is not implemented or provided path is invalid`);
}

export function applyChangesFromServer(manuscript: Manuscript, changesJson: JSONObject[]): Manuscript {
  const allChanges = new BatchChange(deserializeChanges(changesJson));
  return allChanges.applyChange(manuscript);
}
