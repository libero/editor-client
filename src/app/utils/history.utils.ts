import { Transaction } from 'prosemirror-state';

import { ManuscriptHistory } from './state.utils';
import { ManuscriptDiff, ManuscriptDiffValues } from 'app/types/manuscript';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';

export function updateManuscriptState(
  state: ManuscriptHistory,
  propName: string,
  transaction: Transaction
): ManuscriptHistory {
  const change = new ProsemirrorChange(propName, transaction);

  if (transaction.docChanged) {
    return {
      past: [...state.past, change],
      present: change.applyChange(state.present),
      future: []
    } as ManuscriptHistory;
  } else {
    return {
      ...state,
      present: change.applyChange(state.present)
    };
  }
}

export function undoChange(state: ManuscriptHistory): ManuscriptHistory {
  const past = [...state.past];
  const change = past.pop();

  return {
    past,
    present: change.rollbackChange(state.present),
    future: [change, ...state.future]
  };
}

export function createDiff(changes: Record<string, ManuscriptDiffValues>): ManuscriptDiff {
  return {
    ...changes,
    _timestamp: Date.now()
  } as ManuscriptDiff;
}

export function redoChange(state: ManuscriptHistory): ManuscriptHistory {
  const future = [...state.future];
  const change = future.shift();

  return {
    past: [...state.past, change],
    present: change.applyChange(state.present),
    future: future
  };
}
