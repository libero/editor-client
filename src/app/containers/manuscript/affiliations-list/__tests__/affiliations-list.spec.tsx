import React from 'react';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { getInitialHistory, getLoadableStateSuccess } from 'app/utils/state.utils';
import { EditorState } from 'prosemirror-state';
import { create } from 'react-test-renderer';
import { AffiliationsList } from 'app/containers/manuscript/affiliations-list/index';

jest.mock('@material-ui/core', () => ({
  Button: ({ label }) => <div data-cmp="Button">{label}</div>,
  IconButton: () => <div data-cmp="IconButton"></div>
}));

const AFFILIATIONS = [
  {
    label: '1',
    institution: {
      name: 'eLife Sciences',
      department: 'Production Department'
    },
    address: {
      city: 'Cambridge'
    },
    country: 'United Kingdom',
    id: 'aff1'
  },
  {
    label: '2',
    institution: {
      name: 'University',
      department: 'Department'
    },
    address: {
      city: 'City'
    },
    country: 'Country',
    id: 'aff2'
  }
];

describe('Authors details info', () => {
  const mockStore = configureMockStore([]);

  it('renders affiliations list', () => {
    const mockState = getInitialHistory({
      title: new EditorState(),
      abstract: new EditorState(),
      authors: [],
      keywordGroups: {},
      affiliations: AFFILIATIONS
    });
    const store = mockStore({
      manuscript: getLoadableStateSuccess(mockState)
    });
    const wrapper = create(
      <Provider store={store}>
        <AffiliationsList />
      </Provider>
    );

    expect(wrapper).toMatchSnapshot();
  });
});
