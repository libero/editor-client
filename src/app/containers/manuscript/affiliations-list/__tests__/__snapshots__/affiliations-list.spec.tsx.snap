// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors details info renders affiliations list 1`] = `
<section>
  <div
    className="makeStyles-root-7 makeStyles-plainVariant-9"
  >
    <label
      className="makeStyles-label-10"
    >
      Affiliations
    </label>
    <div
      className="makeStyles-listItem-2"
    >
      <div
        className="makeStyles-orderLabel-4"
      >
        (
        1
        )
      </div>
      <div
        className="makeStyles-affiliationInfo-3"
      >
        Production Department, eLife Sciences, Cambridge, United Kingdom
      </div>
      <div
        data-cmp="IconButton"
      />
    </div>
    <div
      className="makeStyles-listItem-2"
    >
      <div
        className="makeStyles-orderLabel-4"
      >
        (
        2
        )
      </div>
      <div
        className="makeStyles-affiliationInfo-3"
      >
        Department, University, City, Country
      </div>
      <div
        data-cmp="IconButton"
      />
    </div>
  </div>
  <div
    data-cmp="Button"
  />
</section>
`;
